Tools/Scripts/libraries/reporelaypy/reporelaypy/checkout.py

```diff
@@ -0,0 +1,201 @@
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import multiprocessing
+import os
+import shutil
+import sys
+
+from webkitcorepy import run
+from webkitscmpy import local
+
+
+class Checkout(object):
+    class Exception(RuntimeError):
+        pass
+
+    class Encoder(json.JSONEncoder):
+        def default(self, obj):
+            if isinstance(obj, dict):
+                return {key: self.default(value) for key, value in obj.items()}
+            if isinstance(obj, list):
+                return [self.default(value) for value in obj]
+            if not isinstance(obj, Checkout):
+                return super(Checkout.Encoder, self).default(obj)
+
+            return dict(
+                path=obj.path,
+                url=obj.url,
+                sentinal=obj.sentinal,
+            )
+
+    @classmethod
+    def from_json(cls, data):
+        data = data if isinstance(data, dict) else json.loads(data)
+        if isinstance(data, list):
+            return [cls(**node, primary=False) for node in data]
+        return cls(**data, primary=False)
+
+    @staticmethod
+    def clone(url, path, sentinal_file=None):
+        run([local.Git.executable(), 'clone', url, path], cwd=os.path.dirname(path))
+        run([local.Git.executable(), 'config', 'pull.ff', 'only'], cwd=path)
+
+        if sentinal_file:
+            with open(sentinal_file, 'w') as cloned:
+                cloned.write('yes\n')
+        return 0
+
+    def __init__(self, path, url=None, http_proxy=None, sentinal=True, primary=True):
+        self.sentinal = sentinal
+        self.path = path
+        self.url = url
+        self._repository = None
+        self._child_process = None
+
+        containing_path = os.path.dirname(path)
+        if not os.path.isdir(containing_path):
+            raise self.Exception("Containing path '{}' does not exist".format(containing_path))
+
+        if http_proxy and run([
+            local.Git.executable(), 'config', '--global', 'http.proxy', 'http://{}'.format(http_proxy),
+        ]).returncode:
+            raise self.Exception("Failed to set https proxy to '{}'".format(http_proxy))
```

Comment:
Do we have concerns about redacting the authentication part of the http_proxy since it technically allows specifying `http://proxyUsername:proxyPassword@proxy.server.com:port`

Reply:
Done